testing/mocks/state-provider/general_history_provider.go

```diff
@@ -0,0 +1,222 @@
+		if end > maxHeightForLevel {
+			return 0, fmt.Errorf(
+				"end %d was greater than max height for level %d",
+				end,
+				maxHeightForLevel,
+			)
+		}
+	}
+	if end < start {
+		return 0, fmt.Errorf("invalid range: end %d was < start %d", end, start)
+	}
+	// The number of hashes is the difference between the start and end
+	// requested heights, plus 1.
+	return (end - start) + 1, nil
+}
+
+// Computes a block history commitment from a start height to a specified
+// height up to the required batch number.
+func (s *L2StateBackend) blockHistoryCommitment(
+	from l2stateprovider.Height,
+	to option.Option[l2stateprovider.Height],
+	batch l2stateprovider.Batch,
+) (commitments.History, error) {
+	var upTo l2stateprovider.Height
+	if !to.IsNone() {
+		upTo = to.Unwrap()
+	} else {
+		blockChallengeLeafHeight := s.challengeLeafHeights[0]
+		upTo = l2stateprovider.Height(blockChallengeLeafHeight)
+	}
+	states, err := s.statesUpTo(uint64(from), uint64(upTo), uint64(batch))
+	if err != nil {
+		return emptyCommit, err
+	}
+	return commitments.New(states)
+}
+
+// Figure out the actual opcode index we should move the machine to
+// when we compute the history commitment. As there are different levels of challenge
+// granularity, we have to do some math to figure out the correct index.
+// Take, for example:
+//
+// lvl2_items_per_lvl1_step = 2
+// lvl3_items_per_lvl2_step = 4
+// lvl4_items_per_lvl3_step = 8
+//
+// This means there are 2 lvl2 items per lvl1 step, 4 lvl3 items per lvl2 step,
+// and 8 lvl4 items per lvl3 step in a challenge.
+//
+// # Let's say we want to compute the actual opcode index for start heights
+//
+// [lvl1_start=2, lvl2_start=3, lvl3_start=4]
+//
+// We can compute the opcode index using the following algorithm for the example above.
+//
+//			2 * (4 * 8) = 64
+//		  + 3 * (8)     = 24
+//	   + 4           = opcode at index 92
+//
+// This generalizes for any number of subchallenge levels into the algorithm below.
+func (s *L2StateBackend) computeMachineStartIndex(
+	startHeights validatedStartHeights,
+) uint64 {
+	if len(startHeights) == 1 {
+		return 0
+	}
+	// We ignore the block challenge level here.
+	heights := startHeights[1:]
+	leafHeights := s.challengeLeafHeights[1:]
+
+	// Next, we compute the opcode index.
+	opcodeIndex := uint64(0)
+	idx := 1
+	// TODO: Handle height 0.
+	for _, height := range heights {
+		total := uint64(1)
+		for i := idx; i < len(leafHeights); i++ {
+			total *= leafHeights[i]
+		}
+		opcodeIndex += total * uint64(height)
+		idx += 1
+	}
+	return opcodeIndex
+}
+
+func (s *L2StateBackend) computeStepIncrement(startHeights validatedStartHeights) (uint64, error) {
+	challengeLevel := len(startHeights) - 1
+	totalChallengeLevels := len(s.challengeLeafHeights)
+
+	// The stepwise increment of the last challenge level is always one.
+	if challengeLevel+1 == totalChallengeLevels {
+		return 1, nil
+	}
+	return s.challengeLeafHeights[challengeLevel+1], nil
```

Comment:
shouldn't this be a multiplication of all challengeLeafHeights from index challengeLevel+1 to the end? just like its done in computeMachineStartIndex ?

Reply:
ah this function tells you in what increments you're supposed to move through the machine. Let's take our current example of block chal, big step, small step

BIG_STEPS_PER_BLOCK_CHAL = 2048
SMALL_STEPS_PER_BIG_STEP = 1,048,576

If Im computing a big step commitment, I want to `machine.Step(BIG_STEPS_PER_BLOCK_CHAL)`. If im doing a small step commitment, I want to do `machine.Step(1)` which is why the edge case above is there. I think it probably needs to change to 

```go
	return s.challengeLeafHeights[challengeLevel], nil
```

Reply:
sorry for the confusion, i should have been more clear, what i meant was let's say we have

BIG_STEPS_PER_BLOCK_CHAL = 5
MEDIUM_STEPS_PER_BIG_STEP = 8
SMALL_STEPS_PER_MEDIUM_STEP = 1,048,576 = 11

then for small step commitment, we wanna do machine.Step(1) and medium step commitment we wanna do machine.Step(1*SMALL_STEPS_PER_MEDIUM_STEP) and for big step commitment we wanna do machine.Step(1*SMALL_STEPS_PER_MEDIUM_STEP*MEDIUM_STEPS_PER_BIG_STEP) right ?

Reply:
Yes you are correct! nice finding